sdk/src/azure/iot/az_iot_pnp_client_twin.c

```diff
@@ -0,0 +1,443 @@
+    size_t* out_mqtt_topic_length)
+{
+  return az_iot_hub_client_twin_document_get_publish_topic(
+      &(client->_internal.iot_hub_client),
+      request_id,
+      mqtt_topic,
+      mqtt_topic_size,
+      out_mqtt_topic_length);
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_parse_received_topic(
+    az_iot_pnp_client const* client,
+    az_span received_topic,
+    az_iot_pnp_client_twin_response* out_response)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  _az_PRECONDITION_NOT_NULL(out_response);
+
+  az_iot_hub_client_twin_response hub_twin_response;
+  _az_RETURN_IF_FAILED(az_iot_hub_client_twin_parse_received_topic(
+      &client->_internal.iot_hub_client, received_topic, &hub_twin_response));
+
+  out_response->request_id = hub_twin_response.request_id;
+  out_response->response_type
+      = (az_iot_pnp_client_twin_response_type)hub_twin_response.response_type;
+  out_response->status = hub_twin_response.status;
+  out_response->version = hub_twin_response.version;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_property_builder_begin_component(
+    az_iot_pnp_client const* client,
+    az_json_writer* ref_json_writer,
+    az_span component_name)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(ref_json_writer);
+  _az_PRECONDITION_VALID_SPAN(component_name, 1, false);
+
+  (void)client;
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(ref_json_writer, component_name));
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+  _az_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(ref_json_writer, component_property_label_name));
+  _az_RETURN_IF_FAILED(
+      az_json_writer_append_string(ref_json_writer, component_property_label_value));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_property_builder_end_component(
+    az_iot_pnp_client const* client,
+    az_json_writer* ref_json_writer)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(ref_json_writer);
+
+  (void)client;
+
+  return az_json_writer_append_end_object(ref_json_writer);
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_begin_property_with_status(
+    az_iot_pnp_client const* client,
+    az_json_writer* ref_json_writer,
+    az_span component_name,
+    az_span property_name,
+    int32_t ack_code,
+    int32_t ack_version,
+    az_span ack_description)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(ref_json_writer);
+  _az_PRECONDITION_VALID_SPAN(property_name, 1, false);
+
+  (void)client;
+
+  if (az_span_size(component_name) != 0)
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(ref_json_writer, component_name));
+    _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+    _az_RETURN_IF_FAILED(
+        az_json_writer_append_property_name(ref_json_writer, component_property_label_name));
```

Comment:
This will emit the component name for every component property

Reply:
discussed offline and will tackle in another PR.